Metadata tag

kalaspuffar opened the pull request:
Issue https://github.com/mtmse/dotify.formatter.impl/issues/25

Description:
Enable the addition of a metadata tag in OBFL to have it available in the PEF output.

Changes:
* Added metadata as a possible child to block.

kalaspuffar commented:
Hi @bertfrees and @PaulRambags

Please review this PR. It is connected to the issue above, and the actual implementation is in the formatter.

Best regards
Daniel

PaulRambags commented on src/obfl-specification.html:
```diff
@@ -1320,6 +1320,29 @@ <h4 id="L836" class="include">page-number</h4>
+<h4 id="L840" class="include">metadata</h4>
+
+<ins datetime="20200429T15:00:00">The metadata attribute can be added to any block in order
+  to send information through the dotify framework from obfl to pef.</ins>
```
Several remarks here.
1. Apparently you have changed the PEF specification. Please make sure that this change is reflected in that specification. See http://pef-format.org/?page_id=7 (is there a "mtmse" copy of this?)
2. This specification has already a section about metadata. See http://braillespecs.github.io/pef/pef-specification.html#Error1. Please make sure that your change to the PEF format is reflected here.
3. Apparently you need metadata in the PEF. Can you use the above mentioned metadata section? In other words, is it _really_ necessary to change the PEF specification?
4. I don't know which organisations are responsible for the PEF format. Have you contacted those organisations?
5. Please don't underestimate the impact of changing the PEF format. For instance, there are companies that produce Braille embossers and their machines accept PEF as input. Changing the PEF format may have a vast impact.

bertfrees replied:
@PaulRambags See https://mtmse.github.io/pef/pef-specification.html

PaulRambags replied:
@kalaspuffar Have you updated https://mtmse.github.io/pef/pef-specification.html?

kalaspuffar replied:
No, we don't change the pef specification.

bertfrees commented on src/obfl-specification.html:
```diff
@@ -1320,6 +1320,29 @@ <h4 id="L836" class="include">page-number</h4>
 
+<h4 id="L840" class="include">metadata</h4>
+
+<ins datetime="20200429T15:00:00">The metadata attribute can be added to any block in order
```
> The metadata attribute

It is an element.

> to any block

I would make this a bit more specific/accurate and say that it can be attached to any position in the text within the block.

bertfrees commented:
@kalaspuffar 

- I think the name `metadata` isn't so great. I explained it here: https://github.com/mtmse/dotify.api/pull/11#issuecomment-621862303
- I'm not so sure about allowing any attribute on the element. I would either allow any attribute _with a foreign namespace_. I think this is more common in XML specifications. Or another idea is to make it very strict, by allowing only one well defined attribute. You could also allow no attributes, but a (foreign) child element. Note that the choice of the exact format is also related to my remark in https://github.com/mtmse/dotify.api/pull/11#issuecomment-621862303 about the type of argument in the Java API.

kalaspuffar commented:
Hi @bertfrees and @PaulRambags 

I've changed the name of the metadata tag to blockmeta to make it more specific and also added the restriction that the attributes need to be in a different namespace than OBFL to be used on the blockmeta tag.

Best regards
Daniel

bertfrees commented:
@kalaspuffar Thanks for the changes.

There is a mistake in the definition of the foreign attribute. It should be like this:

```xml
<attribute>
    <anyName>
        <except>
            <nsName ns=""/>
            <nsName ns="http://www.daisy.org/ns/2011/obfl"/>
        </except>
    </anyName>
</attribute>
```

`blockmeta` is indeed more specific and slightly better, but now it appears like some piece of information that is attached to a block, which it isn't. I would argue to get rid of "meta" in the name altogether because it distracts too much from the essence of the new element. It is mainly about the _position_ of the element. A position to which some metadata could be attached, but that part is secondary.

I'm sorry, I know I should be giving useful name suggestions instead.

What about `bookmark`? Just came to mind.

bertfrees commented:
By the way it can be a good idea to update the "src/examples/example1.obfl" file if you make a change to the schema, because then you can open that file in your editor, with the schema applied. I sometimes do that. It helps to find bugs in the schema.

kalaspuffar commented:
Hi @bertfrees and @PaulRambags 

```bookmark``` sounds good, other suggestions ```navpoint``` from daisy epub. Or perhaps we could use ```ExternalReference``` .

Names are always hard :)

Best regards
Daniel

bertfrees commented:
`navpoint` is also not bad, more or less equivalent to `bookmark`. The connotation with NCX's `navPoint` is maybe not so great though because that is a different beast (it's basically a TOC entry).

`external-reference` is another angle but also good because it highlights that it does not have a meaning within the OBFL document itself, that a link to something outside the OBFL is assumed.

bertfrees commented:
OK so you decided on `external-reference`? OK, I guess I can live with that.

I was still undecided. The "reference" in "external-reference" suggests an _outward_ link while the feature could as well be used the other way around. On the other hand "bookmark" is also a bit suggestive because when you say bookmark you're thinking about something used for _navigating_ to, while there is no such restriction.

Yesterday I was thinking about maybe just using `span`. This is possible if you are using foreign attributes like we do. You are just attaching some custom information (the foreign attributes) to a certain part of the OBFL document. Apart from being short (much shorter than "external-reference") and semantically neutral, it also has the benefit that it is not limited to a single point. With a span you can either mark a single point (empty span) or a segment between two points. And what's more is that you could even extend it to other elements in the future: blocks, sequences, maybe even pages and volumes? It might seem far-fetched now, but it might turn out useful some time if you could see in the output the position and size/shape of blocks and other structural elements like page areas and such. I haven't figured out what the API would look like on the other side though (the writer side). Providing info attached to some arbitrary area is a bit awkward if at the same time you need to indicate where the rows, pages, etc. start and end.

Anyway, I'm probably bike-shedding now. I just like to spend a decent amount of time on thinking things through properly when you work on a specification and good names are an important part of that. But the solution we have now is pretty good already.

kalaspuffar commented:
Hi @bertfrees 

Well, decided it is a lite bit too definitive. :)

But ```external-reference``` is the best suggestion we have so far. And yes, it's a bit long, but it describes how it might be used. Adding spans might not be good, as this might be something we want to use in the OBFL specification natively. I wanted it to be a special tag that told us that it was not a part of the normal document flow. But I'm still open to suggestions.

I will continue refining the formatter implementation, and perhaps another better suggestion is found later. I just like to make progress.

Best regards
Daniel

bertfrees commented:
IMO the special tag is only really needed if you want to enforce (and emphasize) that it is a single point ("atomic" if you will). If you are willing to generalize it to not just a single position, you might as well use span. The fact that we're using foreign attributes already makes it clear that this is some deviation from standard OBFL. I would consider this element a part of the normal document flow though. It's just an invisible element that flows along with other inline elements, very similar to `anchor` or `marker`, or an empty `span` with an `id` attribute. It's some piece of information attached to something. That something can be a single point, but by extension also a visible element.

But anyway, I think we can go for the special empty element. Its attributes will still be extendable to any element. The empty element might then become obsolete one day, but that's not an issue.

kalaspuffar commented:
Hi @bertfrees and @PaulRambags

I'm still waiting for review and approval for this PR.

Best regards
Daniel

PaulRambags commented on src/obfl-specification.html:
```diff
@@ -1167,7 +1167,7 @@ <h4 id="L1023" class="include">block</h4>
     </dd>
   <dt>Content Model</dt>
-    <dd>text(), leader, marker, br, evaluate, page-number, marker-reference,
+    <dd>text(), leader, marker, br, evaluate, page-number, blockmeta, marker-reference,
```
`blockmeta`? I would expect `external-reference`

kalaspuffar replied:
Well, I forgot to update the specification. Thanks for notifying me.

PaulRambags commented on src/obfl-specification.html:
```diff
@@ -1320,6 +1320,29 @@ <h4 id="L836" class="include">page-number</h4>
 </dl>
 </div>
 
+<h4 id="L840" class="include">blockmeta</h4>
```
`external-reference`

PaulRambags commented on src/obfl-specification.html:
```diff
@@ -1167,7 +1167,7 @@ <h4 id="L1023" class="include">block</h4>
     </dd>
   <dt>Content Model</dt>
-    <dd>text(), leader, marker, br, evaluate, page-number, marker-reference,
+    <dd>text(), leader, marker, br, evaluate, page-number, blockmeta, marker-reference,
```
The `text()`, `leader`, `marker`, `br`, etc. are so-called "block contents". If you allow an `external-reference` here, you must also allow it _everywhere_. That is, all these elements should have `external-reference` in their Content Model:
```
before
page-area
block
td
toc-entry
toc-entry-on-resumed
volume-transition
block-resumed
item
```

kalaspuffar replied:
No, the content model for a block don't have to be equal to other similar types of elements that are block-like. Having external-references in a table of contents is not necessarily something that we want.

Then again for use-cases that @bertfrees have pointed out, it could be something we want to add in the future.

bertfrees replied:
@kalaspuffar is right, however I think @PaulRambags has a good point that it would bring consistency in the spec if you also change the content model of these other elements. You don't even need to support it in Dotify at this point, although it would be trivial to do (if it's not already the case).

bertfrees replied:
To extend it to elements like `field` is also a possibility although a slightly different story because it might require some more work in the implementation.

kalaspuffar replied:
Added to the content model of blocks below.
```
before
block
td
toc-entry
toc-entry-on-resumed
block-resumed
item
field
```

I did not add it to these because they are ordered in a specific order which and the reference doesn't have a good place to be in the order.
```
volume-transition
page-area
```

bertfrees replied:
> I did not add it to these because they are ordered in a specific order which and the reference doesn't have a good place to be in the order.

What do you mean exactly?

Oh by the way there is also `after`, which Paul didn't mention. And there might be other elements, I haven't checked.

kalaspuffar replied:
Added external-reference on all places we use marker-reference.

What I meant was that page-area and volume-transition as elements don't need a direct child of external-reference. They have particular children in order. Some of the children can have references so you should be able to put the reference in at a suggested position.

bertfrees replied:
Ah yes, of course.

bertfrees commented on src/obfl-specification.html:
```diff
@@ -1320,6 +1320,29 @@ <h4 id="L836" class="include">page-number</h4>
+<h4 id="L840" class="include">external-reference</h4>
+
+<ins datetime="20200429T15:00:00">The external-reference element can be attached to any position in the text within the block
+  to send information through the dotify framework from obfl to pef.</ins>
```
Hmm, this might seem nitpicking, but it feels wrong to mention Dotify here. There are no references to Dotify in the whole document. Replace it with for instance "user agent" or "implementing application" (see the "General Terms" section).

Same story for PEF, although here it might actually help to explain what the feature is about. But then there should be a reference in the "References" section at the bottom.

What about the following:

> With the external-reference element custom annotations can be attached to any position in the text within a block or field. How the user agent processes the information is implementation dependent.
> 
> ### Example
> 
> A typical use case is to pass information through unchanged from the source document to the destination, for instance a PEF document [PEF]. The result document is annotated with the exact same information attached to the corresponding positions in the braille.
> 
> _then some example XML in- and output:_
>
> ```xml
> <obfl> ... </obfl>
> ```
> ```xml
> <pef> ... </pef>
> ```
>
> _and below, under References_:
>
> **[PEF]**
>     "[PEF 1.0 - Portable Embosser Format (Public draft)](https://mtmse.github.io/pef/pef-specification.html)", Joel Håkansson, 12 September 2011.
>     [Latest version](https://mtmse.github.io/pef/) available at: https://mtmse.github.io/pef

kalaspuffar replied:
Hi @bertfrees

Well, writing a specification requires some nitpicking :)

I've added the reference in the specification and some examples and changed to the wording you suggested. It seemed like a really good suggestion.

Best regards
Daniel

PaulRambags commented on src/obfl-specification.html:
```diff
@@ -1543,7 +1593,7 @@ <h4 id="L2277" class="include">td</h4>
     </dd>
   <dt>Content Model</dt>
-  <dd>text(), leader, marker, br, evaluate, page-number, marker-reference, span,
+  <dd>text(), leader, marker, br, evaluate, page-number, marker-reference, span, external-reference,
```
You are inconsistent in the order of the block attributes.
Everywhere above this line the `external-reference` follows the `page-number` element and here it follows the `span` element. To increase readability and to show that all these block attributes are identical everywhere, please make sure that the order of the block attributes is the same accross the whole document.

kalaspuffar replied:
Re-ordered attributes.
